src/rust/engine/src/externs.rs

```diff
@@ -509,11 +494,43 @@ pub enum GeneratorResponse {
 /// the object's type.
 ///
 #[repr(C)]
+#[derive(Clone, Copy)]
 pub struct Ident {
   pub hash: i64,
   pub type_id: TypeId,
 }
 
+pub trait RawBuffer<Raw, Output> {
+  fn ptr(&self) -> *mut Raw;
+  fn len(&self) -> u64;
+
+  ///
+  /// A buffer-specific shallow clone operation (possibly just implemented via clone).
+  ///
+  fn lift(t: &Raw) -> Output;
+
+  ///
+  /// Returns a Vec copy of the buffer contents.
+  ///
+  fn to_vec(&self) -> Vec<Output> {
+    with_vec(self.ptr(), self.len() as usize, |vec| {
+      vec.iter().map(Self::lift).collect()
+    })
+  }
+
+  ///
+  /// Asserts that the buffer contains one item, and returns a copy of it.
+  ///
+  fn unwrap_one(&self) -> Output {
+    assert!(
+      self.len() == 1,
+      "Expected exactly 1 item in Buffer, but had: {}",
+      self.len()
+    );
+    with_vec(self.ptr(), self.len() as usize, |vec| Self::lift(&vec[0]))
```

Comment:
Any reason this doesn't delegate to `.to_vec()`? Like:
```rust
  fn unwrap_one(&self) -> Output {
    assert_equal!(self.len(), 1, "Expected exactly 1 item in Buffer, but had: {}", self.len());
    self.to_vec()[0]
  }
```

Reply:
`with_vec` doesn't actually "create" a vec: instead, it "views" some memory as a Vec via:
```
unsafe { Vec::from_raw_parts(c_ptr, c_len, c_len) }
```
That means that by avoiding calling `to_vec()`, we avoid actually copying the buffer containing the one item.